kythe/go/platform/kcd/kzipdb/kzipdb_test.go

```diff
@@ -0,0 +1,198 @@
+}
+
+func setup(t *testing.T) {
+	t.Helper()
+	testInput.Do(func() {
+
+		buf := bytes.NewBuffer(nil)
+		w, err := kzip.NewWriter(buf)
+		if err != nil {
+			t.Fatal("NewWriter: %v", err)
+		}
+		for _, ic := range []*apb.IndexedCompilation{
+			{Unit: newUnit("A", "go", "kythe"), Index: newIndex("123")},
+			{Unit: newUnit("B", "c++", "kythe"), Index: newIndex("456", "789")},
+			{Unit: newUnit("C", "go", "chromium"), Index: newIndex("123")},
+			{Unit: newUnit("D", "protobuf", "kythe"), Index: newIndex("789")},
+		} {
+			digest, err := w.AddUnit(ic.Unit, ic.Index)
+			if err != nil {
+				t.Fatalf("AddUnit %v: %v", ic.Unit, err)
+			}
+			testInput.units = append(testInput.units, digest)
+		}
+		for _, src := range []string{
+			"apple", "pear", "plum", "cherry",
+		} {
+			digest, err := w.AddFile(strings.NewReader(src))
+			if err != nil {
+				t.Fatalf("AddFile %q: %v", src, err)
+			}
+			testInput.files = append(testInput.files, digest)
+		}
+		if err := w.Close(); err != nil {
+			t.Fatalf("Close: %v", err)
+		}
+
+		testInput.data = buf.Bytes()
+	})
+}
+
+func TestReader(t *testing.T) {
+	setup(t)
+	r, err := kzip.NewReader(bytes.NewReader(testInput.data), int64(len(testInput.data)))
+	if err != nil {
+		t.Fatalf("NewReader failed: %v", err)
+	}
+	db := DB{Reader: r}
+	ctx := context.Background()
+
+	t.Run("NoMatchingRevs", func(t *testing.T) {
+		db.Revisions(ctx, &kcd.RevisionsFilter{Corpus: "bogus"}, func(rev kcd.Revision) error {
+			t.Errorf("Unexpected revision found: %v", rev)
+			return nil
+		})
+	})
+	t.Run("MatchSomeRevs", func(t *testing.T) {
+		want := []kcd.Revision{
+			{Revision: "123", Corpus: "kythe"},
+			{Revision: "456", Corpus: "kythe"},
+			{Revision: "789", Corpus: "kythe"},
+		}
+		var got []kcd.Revision
+		db.Revisions(ctx, &kcd.RevisionsFilter{Corpus: "kythe"}, func(rev kcd.Revision) error {
+			got = append(got, rev)
+			return nil
+		})
+		sort.Slice(got, func(i, j int) bool {
+			if c := strings.Compare(got[i].Revision, got[j].Revision); c == 0 {
+				return got[i].Corpus < got[j].Corpus
+			} else {
+				return c < 0
+			}
+		})
+		if diff := cmp.Diff(got, want); diff != "" {
+			t.Errorf("Revisions did not match: diff is\n%s", diff)
+		}
+	})
+	t.Run("FindGo", func(t *testing.T) {
+		var numFound int
+		db.Find(ctx, &kcd.FindFilter{Languages: []string{"go"}}, func(digest string) error {
+			unit, err := r.Lookup(digest)
+			if err != nil {
+				t.Errorf("Lookup %q failed: %v", digest, err)
+			} else if lang := unit.Proto.GetVName().GetLanguage(); lang != "go" {
+				t.Errorf("Language for %q: got %q, wanted go", digest, lang)
+			} else {
+				numFound++
+			}
+			return nil
+		})
+		if numFound != 2 {
+			t.Errorf("Find: got %d compilations, want 2", numFound)
+		}
+	})
+	t.Run("ReadUnits", func(t *testing.T) {
+		i := 0
+		db.Units(ctx, testInput.units, func(digest, key string, data []byte) error {
+			if want := testInput.units[i]; digest != want {
```

Comment:
Is it safe to assume the order of processing will be the same as the test input array order?

Reply:
In this implementation yes. It's not a required property of the interface, but since this test is specific to this format it's safe.